Objects/asm_trampoline.S

```diff
@@ -0,0 +1,28 @@
+    .text
+    .globl	_Py_trampoline_func_start
+# The following assembly is equivalent to:
+# PyObject *
+# trampoline(PyThreadState *ts, _PyInterpreterFrame *f,
+#            int throwflag, py_evaluator evaluator)
+# {
+#     return evaluator(ts, f, throwflag);
+# }
+_Py_trampoline_func_start:
+#ifdef __x86_64__
+    sub    $8, %rsp
+    call    *%rcx
+    add    $8, %rsp
+    ret
```

Comment:
As we experimented, another possible implementation is 
```
    pushq  %rbp
    movq   %rsp, %rbp
    callq  *%rcx
    popq   %rbp
    retq
```
that consumes slightly smaller space (8 bytes) than the one here (11 bytes) and coincidentally has a good memory alignment.

Reply:
I started with that one but unfortunately, gdb and other unwinders don't unwind across this version (all checks are done with `CFLAGS="-fno-omit-frame-pointer -mno-omit-leaf-frame-pointer"`):

```
#0  0x000055555571178f in pymalloc_alloc (_unused_ctx=<optimized out>, nbytes=264) at Objects/obmalloc.c:1980
#1  _PyObject_Malloc (ctx=<optimized out>, nbytes=264) at Objects/obmalloc.c:1999
#2  0x00005555556fe758 in new_values (size=29) at Objects/dictobject.c:699
#3  init_inline_values (tp=<optimized out>, obj=0x7ffff74e4fe0) at Objects/dictobject.c:5323
#4  _PyObject_InitializeDict (obj=obj@entry=0x7ffff74e4fe0) at Objects/dictobject.c:5346
#5  0x0000555555722c3f in object_new (type=<optimized out>, args=<optimized out>, kwds=<optimized out>)
    at Objects/typeobject.c:4907
#6  0x0000555555724ba6 in type_call (type=0x555555dc0290, args=0x555555ad00e0 <_PyRuntime+68256>, kwds=0x0)
    at Objects/typeobject.c:1199
#7  0x00005555556b02e0 in _PyObject_MakeTpCall (tstate=0x555555aeb758 <_PyRuntime+180504>,
    callable=callable@entry=0x555555dc0290, args=args@entry=0x7ffff7f9e068, nargs=<optimized out>,
    keywords=keywords@entry=0x0) at Objects/call.c:215
#8  0x00005555556b110d in _PyObject_VectorcallTstate (kwnames=0x0, nargsf=<optimized out>, args=0x7ffff7f9e068,
    callable=0x555555dc0290, tstate=<optimized out>) at ./Include/internal/pycore_call.h:90
#9  0x000055555564bb70 in _PyEval_EvalFrameDefault (tstate=0x555555aeb758 <_PyRuntime+180504>, frame=<optimized out>,
    throwflag=<optimized out>) at Python/ceval.c:4274
#10 0x00007ffff7b76bbe in ?? ()
#11 0x00007fffffff98d0 in ?? ()
#12 0x0000555555788d48 in py_trampoline_evaluator (ts=0x555555aeb758 <_PyRuntime+180504>, frame=0x7ffff7f9dfd0, throw=0)
    at Objects/perf_trampoline.c:384

Backtrace stopped: frame did not save the PC
```

but it does across the current one:

```
#0  0x00005555556e4181 in long_to_decimal_string_internal (aa=0x555555abf990 <_PyRuntime+848>,
    p_output=p_output@entry=0x7fffffff9400, writer=writer@entry=0x0, bytes_writer=bytes_writer@entry=0x0,
    bytes_str=bytes_str@entry=0x0) at Objects/longobject.c:1742
#1  0x00005555556e6527 in long_to_decimal_string (aa=<optimized out>) at Objects/longobject.c:1920
#2  0x000055555570c165 in PyObject_Repr (v=0x555555abf990 <_PyRuntime+848>) at ./Include/object.h:135
#3  0x00005555556dc80d in list_repr (v=0x7ffff722cdc0) at Objects/listobject.c:405
#4  list_repr (v=0x7ffff722cdc0) at Objects/listobject.c:374
#5  0x000055555570c165 in PyObject_Repr (v=0x7ffff722cdc0) at ./Include/object.h:135
#6  0x0000555555708abb in cfunction_vectorcall_O (func=0x7ffff7c58770, args=0x7ffff7f9e0b8, nargsf=<optimized out>,
    kwnames=<optimized out>) at Objects/methodobject.c:514
#7  0x00005555556b10cd in _PyObject_VectorcallTstate (kwnames=0x7fffffff94b0, nargsf=<optimized out>,
    args=0xfffffffff7fa4c80, callable=0x7ffff7c58770, tstate=0x555555aeb758 <_PyRuntime+180504>)
    at ./Include/internal/pycore_call.h:92
#8  PyObject_Vectorcall (callable=callable@entry=0x7ffff7c58770, args=args@entry=0x7ffff7f9e0b8, nargsf=<optimized out>,
    kwnames=kwnames@entry=0x0) at Objects/call.c:300
#9  0x000055555564bb70 in _PyEval_EvalFrameDefault (tstate=0x555555aeb758 <_PyRuntime+180504>, frame=<optimized out>,
    throwflag=<optimized out>) at Python/ceval.c:4274
#10 0x00007ffff7b77b89 in ?? ()
#11 0x00007fffffff96b0 in ?? ()
#12 0x0000555555788d48 in py_trampoline_evaluator (ts=0x555555aeb758 <_PyRuntime+180504>, frame=0x7ffff7f9e040, throw=0)
    at Objects/perf_trampoline.c:384
#13 0x00005555557c1b3b in _PyEval_EvalFrame (throwflag=0, frame=0x7ffff7f9e040, tstate=0x555555aeb758 <_PyRuntime+180504>)
```

Reply:
Thanks for detailed explanation! That is an interesting problem and I also saw the same error message ("Backtrace stopped: frame did not save the PC") when test code with gdb. However, other tools, such as lldb debugger and absl unwinder ([absl:: GetStackTraceWithContext](https://github.com/abseil/abseil-cpp/blob/master/absl/debugging/stacktrace.cc#L108)), worked for our use case, so it seemed to be related to the the unwinding logic.